### ISSUE TITLE
Rework Validation Framework?

### ISSUE THREAD

**Issue by frizbog** [frizbog/gedcom4j#48]
Now that most objects in the data model extend AbstractElement, it might make sense to add an abstract validate() method to that class and move validation code into the model itself.  If each object knows how to validate itself, it might be cleaner??

Of course, that would make the objects in the model more than simple value objects.  

Need to think about this.

**Comment by frizbog** [comment]
I've started working on this in v3-development branch.

**Comment by frizbog** [comment]
After careful consideration I've abandoned the Visitor pattern idea for validation and am keeping the existing Decorator-based validation. While it's cool, there are several drawbacks:

1. Visitor pattern requires adding an accept method to each model class so they can be visited, which eroded at a design goal of have the model classes be value objects without outbound hooks.  The interface for visitors still shields the object model classes from the particulars of what's visiting it, but I still didn't like adding that method to every class.
2. All the double-dispatch involved slows things down and impacts performance.
3. There are so many objects in the object model the concrete visitors themselves immediately require 40+ visit methods to be implemented.  This encourages creating long "god classes” every time a visitor is implemented.
4. Visitor pattern is most called-for when the entire object graph is expected to be visited in new ways over time…but there was only one Visitor ever needed by the library, for validation.  That’s a lot of complication and hassle and unpleasantness for a single use case. I did consider whether consumers of the library would find support for Visitor useful, and decided that consumers would generally prefer to navigate directly to the classes of interest (which are easy to get to), rather than to obligate themselves to implementing a Visitor that keeps object graph traversal going.
